### ISSUE TITLE
everytime out_z is zero 

### ISSUE THREAD

**Issue by havvanurbozomeroglu** [plemeri/ros_vision_pkg#1]
Hi, 

 ```
   max_deg = int(360 // sample_rate)
    arr = pointcloud2_to_xyz_array(msg)

    angle = np.arctan2(arr[:, 1], arr[:, 0])
    angle = np.rad2deg(angle) + 180
    angle /= sample_rate
    angle = np.round(angle)
    angle[angle > max_deg - 1] = max_deg - 1

    dist = np.sqrt(arr[:, 0] ** 2 + arr[:, 1] ** 2)
    min_dist = np.ones(max_deg) * max_dist
    for i, ang in enumerate(angle):
        if min_dist[int(ang)] > dist[i]:
            min_dist[int(ang)] = dist[i]
    # -- make angle based min_distance data -- end

    # -- to pointcloud for visualization
    out_angle = np.deg2rad(np.linspace(0, 359, max_deg))
    out_x = -np.cos(out_angle) * min_dist
    out_y = -np.sin(out_angle) * min_dist
    out_z = np.zeros(max_deg)
    out_cloud = np.stack([out_x, out_y, out_z], axis=-1)

    print(out_z)
```
output: 
`[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
[0. 0. 0. 0. 0. 0. 0.]
`

print(out_x)

output: 
`[-0.         -0.30125702  0.30265708  0.99996192  0.         -0.30468984
 -0.        ]
[-0.         -0.30637342  0.28919702  0.99996192  0.         -0.30468984
 -0.        ]
[-0.         -0.31355883  0.30099976  0.99996192  0.         -0.30612661
 -0.        ]
[-0.         -0.30135081  0.28919702  0.99996192  0.         -0.29824787
 -0.        ]`

ı use ouster os1. Thanks in advance

**Comment by plemeri** [comment]
Hello, we currently do not support any public use for this repository since we this for Autonomous Vehicle Challenge in Korea, but you found our project anyhow and happy to be your help! 

First, the source code that you commented seems from the [`points_ring`](https://github.com/plemeri/ros_vision_pkg/tree/main/src/points_ring) package which should be used with [`ray_ground_filter`](https://github.com/plemeri/ros_vision_pkg/tree/main/src/ray_ground_filter) to publish `\points_no_ground`.

This is because we made points_ring package to utilize pointcloud data as a collision detector. It's a distance from the nearest obstacle from 0 to 359 degrees. Normally, it is very easy to use ground removed pointcloud as is, but some of my colleague had a hard time understanding the data-format, transform in ROS, etc, so I made it very simple, so anyone can understand.

If you need still need this format, we made another version [`freespace_detector`](https://github.com/plemeri/ros_vision_pkg/tree/main/src/freespace_detector) which is now embedded into one launch file. This time, the output of points_ring not pointcloud2, rather, it is MarkerArray and the topic name is `\marker_ground` instead which you can easily change from launch file.

Long story short, please let me know if you are having any trouble with this repository. I'm happy to help.

**Comment by havvanurbozomeroglu** [comment]
Hi @plemeri ,
First of all, thank you very much for your answer. I'm having a hard time understanding it as I'm new to both ros and lidar. I have an ouster lidar and am trying to use it in a self-driving vehicle. I want to get the distances from the obstacles around the vehicle with the lidar. I've been trying for a week but I can't find the result. I would be very grateful if you could help me with this. What exactly do I need to use?

I want to get range values this way if possible. I completely made up the number values here.
regions= {
        'right':  min(min(data.ranges[0:143]), 30),
        'front_right': min(min(data.ranges[144:287]), 30),
        'front':  min(min(data.ranges[288:431]), 30),
        'front_left':  min(min(data.ranges[432:575]), 30),
        'left':   min(min(data.ranges[576:719]), 30)
    }

I would be very happy if you guide me. Thank you very much in advance.

**Comment by plemeri** [comment]
In that case, I would recommend using `freespace_detector`. Here, let me make it more simple to use for you.
Create a new launch file and copy the below script.
```
<launch>
    <arg name="pointcloud_topic"    default="/os_cloud_node/points" />
    <arg name="freespace_topic"     default="/os_cloud_node/freespace" />
    <arg name="freespace_vis_topic" default="/os_cloud_node/freespace_vis" />
    
    <include file="$(find freespace_detector)/launch/livox_free_space.launch">
        <arg name="pointcloud_topic"  value="$(arg pointcloud_topic)" />
        <arg name="result_topic"      value="$(arg freespace_topic)" />
        <arg name="vis_topic"         value="$(arg freespace_vis_topic)" />
    </include>
</launch>
```

I assume that you are familiar with launch file. Copy above script and just plug your `PointCloud2` topic.
You have to subscribe `freespace_vis_topic` which is `MarkerArray`. Please refer to the official documentation of `MarkerArray`.

You can then subscribe each `Marker` from the `MarkerArray`, where each `Marker` contains positional information as `marker.pose.position` and each angle information as `marker.id`. Note that angle value is in degree with integer type.
Once you subscribe all markers, you'll have a position, or distance for each degree (0 ~ 359). 

<img width="1072" alt="Screen Shot 2022-06-09 at 9 58 39 PM" src="https://user-images.githubusercontent.com/40786892/172854798-0d5a8d01-c475-48e9-bef9-64f94869b668.png">

Here's how the `freespace_vis_topic` visualizes in `rviz`. The green `Marker` denotes 0°, and it increases clockwise.
In your case, you can first just subscribe all `Marker`s to the list, and slice it out for certain range of degrees that you desire. For instance, if you successfully subscribed `freespace_vis_topic` into a list called `dist`, then the right region for 30° range would be `dist[-15:] + dist[:15]`. Be careful with the 0° since it does not start from the front :)

Hope this can help with your problem!

**Comment by havvanurbozomeroglu** [comment]
`freespace_vis_topic` visualizes in rviz. where am i doing wrong? @plemeri 
`dist` values are also abnormal.
thanks a lot for your answers

![github2](https://user-images.githubusercontent.com/59619952/172935114-2ead5153-c584-4763-baa5-ab1d82d84d53.png)

1. run ouster_launch 
2. run livox_free_space.launch
3. run start.launch

start.launch
```

<launch>
    <arg name="pointcloud_topic"    default="/os_cloud_node/points" />
    <arg name="freespace_topic"     default="/os_cloud_node/freespace" />
    <arg name="freespace_vis_topic" default="/os_cloud_node/freespace_vis" />
    
    <include file="$(find freespace_detector)/launch/livox_free_space.launch">
        <arg name="pointcloud_topic"  value="$(arg pointcloud_topic)" />
        <arg name="result_topic"      value="$(arg freespace_topic)" />
        <arg name="vis_topic"         value="$(arg freespace_vis_topic)" />
    </include>
</launch>
```

livox_free_space.launch
```

<launch>

    <arg name = "height_offset"    default="1.8"/>
    <arg name = "pointcloud_topic" default="/os_cloud_node/points"/>
    <arg name = "result_topic" default="/points_ground"/>
    <arg name = "vis_topic" default="/marker_ground"/>
    <arg name = "sample_rate" default="1.0"/>

    <node pkg="freespace_detector" name="freespace_detector" type="freespace_detector_node" output="screen"/>
        <param name = "height_offset"    type="double" value="$(arg height_offset)"/>
        <param name = "pointcloud_topic" type="str"    value="$(arg pointcloud_topic)"/>
        <param name = "result_topic"     type="str"    value="$(arg result_topic)"/>

    <node name="freespace_parser" pkg="freespace_detector" type="livox_parser.py">
        <param name ="height_offset"   type="double" value="$(arg height_offset)"/>
        <param name ="freespace_topic" type="str"    value="$(arg result_topic)" />
        <param name ="result_topic"    type="str"    value="$(arg vis_topic)" />
        <param name ="sample_rate"     type="double" value="$(arg sample_rate)" />
    </node>

</launch>
```

thank you for taking the time :)

**Comment by plemeri** [comment]
You don't need to launch two of them at the same time. Just launch `start.launch`.

**Comment by BehiceB** [comment]
hello @plemeri,
in this project i got dist values but i couldn't use it. which ranges i can use for usage? can you help me about this?

**Comment by plemeri** [comment]
Could you be more specific? If you subscribed in a right way, any range can be used.

**Comment by havvanurbozomeroglu** [comment]
dist[-15:] +dist[15:] gives 30°. We don't understand what numbers we need to use to get between 30 and 60, we try something, but it seems like the wrong data.  In the driverless vehicle, we want to divide the 180-degree angle into intervals and obtain it.

**Comment by plemeri** [comment]
I think I cannot help you more for now. Can you send me your pointcloud data by recording bagfile with `rosbag`?
You can easily record your data by following command,
```
rosbag record -0 test.bag /os_cloud_node/points
```
Just 5-10 seconds amount would be enough. Press `ctrl+c` to end recording, and send me the `test.bag` file via e-mail.
Here's my e-mail address. taehoon1018@postech.ac.kr

I'm in the middle of something right now, so I cannot be sure that I can find the problem right away, but I'll try my best.

**Comment by havvanurbozomeroglu** [comment]
hello @plemeri, we sent you an email about test.bag file. Could you have any time to examine this?

**Comment by plemeri** [comment]
Here's what I got from you.
<img width="874" alt="Screen Shot 2022-06-16 at 1 50 50 PM" src="https://user-images.githubusercontent.com/40786892/173993742-9282df9d-6f41-45f9-9ff2-8d6286092a50.png">
See, unlike usual pointcloud data, I found that yours don't have a 'ring shaped' flat surface, which is a key component for this package.
Long story short, I don't think this is the best option depending on your recorded bagfile.

Instead, I recommend using `ray_ground_filter`. Use following launch file instead.
```
<launch>
    <arg name="base_frame"          default="os_sensor" />
    <arg name="pointcloud_topic"    default="/os_cloud_node/points" />
    <arg name="freespace_topic"     default="/os_cloud_node/freespace" />
    <arg name="freespace_vis_topic" default="/os_cloud_node/freespace_vis" />
    <arg name="local_max_slope"     default="8" />
    <arg name="general_max_slope"   default="5" />
    <arg name="min_point_distance" default="1.85" /> 
    

    <include file="$(find points_preprocessor)/launch/ray_ground_filter.launch">
        <arg name="base_frame"            value="$(arg base_frame)" />
        <arg name="input_point_topic"     value="$(arg pointcloud_topic)" />
        <arg name="no_ground_point_topic" value="/points_no_ground" />
        <arg name="local_max_slope"       value="$(arg local_max_slope)" />
        <arg name="general_max_slope"     value="$(arg general_max_slope)" />
        <arg name="min_point_distance"     value="$(arg min_point_distance)" />
    </include>

    <include file="$(find points_ring)/launch/points_ring.launch">
        <arg name="input_topic"       value="/points_no_ground" />
        <arg name="output_topic"      value="$(arg freespace_topic)" />
        <arg name="output_vis_topic"  value="$(arg freespace_vis_topic)" />
    </include>

</launch>
```
Results are like this.
<img width="902" alt="Screen Shot 2022-06-16 at 5 43 15 PM" src="https://user-images.githubusercontent.com/40786892/174030968-2d201bc4-9d47-4413-9d63-cddf90971f70.png">

Note that the format is different from `livox_free_space`.
The desired topic name is same as `/os_cloud_node/freespace`, but in this case it's `numpy_msg` in `points_ring` package. If you don't have much knowledge of custom messages in ROS, just modify `points_ring.py` and access to the data directly.

Good luck.
